### ISSUE TITLE
No fermionic CTM energy path is asserted against a reference (the ED machinery exists but only guards the contractor path)

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#879]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @yingjerkao.

**No fermionic CTM energy path is asserted against a reference.** The ED machinery to do it already exists in the repo and is used on exactly one path — the contractor reference — while the paths `fpeps()` actually returns are compared to it by `print()` and nothing else.

## What exists, precisely

`tests/test_fermionic_ed_reference.py` builds a real reference: `_build_tv_hamiltonian_2x2_pbc(t, V)` plus `np.linalg.eigvalsh`, giving a 4-site 2×2 PBC Jordan-Wigner ED ground state. Measured: **E_gs/site = −0.5000000000** at t=1 for both V=0 and V=0.5.

`test_tier7_variational_bound_check` uses it properly — 7 seeds, asserting `e_ref >= E_gs_per_site - 1e-9`, and failing on violations. That is a genuine physics check.

But it validates **`reference_energy_2x2_pbc`**, the direct contractor path. It does not touch the CTM.

## The gap

`test_tier5_compare_to_ipeps_ctm_energy` is the only place the CTM energies meet the reference, and it **asserts nothing**:

```python
    print(f"\n[Tier 5] reference E (2x2 PBC, contractor):  {e_ref}")
    print(f"[Tier 5] iPEPS-CTM E (infinite, chi={chi}):    {e_ctm:.10f}")
    print(f"[Tier 5] split-CTM E (via shim fallback):     {e_split:.10f}")
    print(f"[Tier 5] split-CTM E (raw bar_super path):    {e_split_raw:.10f}")
    print(f"[Tier 5] |e_ref - e_ctm|:    {abs(e_ref - e_ctm):.3e}")
```

Its docstring is explicit that this is deliberate — *"These are NOT expected to match exactly … We print both to anchor a baseline"* — which is defensible for an infinite-lattice-vs-2×2-torus comparison. The problem is that nothing else covers it either. Every other fermionic energy assertion in the tree is a finiteness, positivity or shape check:

| test | asserts |
|---|---|
| `test_tier5_reference_energy_finite` | `np.isfinite` |
| `test_tier6_jw_ed_ground_state_finite` | `np.isfinite`, Hermiticity |
| `test_tier8_obc_norm_is_positive` | positivity |
| `test_fermionic_ctm_split_2site_public_path` | `jnp.isfinite(E)`, `|Im E| < 1e-8` |

So `compute_energy_ctm_tensor`, `compute_energy_split_ctm_tensor` and `compute_energy_split_ctm_tensor_2site` can each return any finite real number on a fermionic state and no test objects.

## Why it matters now

While fixing #878 I moved `fpeps()` to a 2-site checkerboard. On a state whose bond spectrum is healthy (`lam_h = [1, 0.678]` at D=2, 100 steps) and whose corner is full rank with no collapse warning, `compute_energy_split_ctm_tensor_2site` returns **≈ −5e−5**.

I cannot tell whether that is a defect I introduced or one that was always there, **because there is no assertion anywhere that would have caught either**. That is the same shape as #878 itself: `fpeps()` returned exactly zero for months because `|A|` was renormalised to 1.0 on the way down and every `isfinite` check passed on the corpse.

## Suggested fix

The cheap version is already written — `test_tier7_variational_bound_check`'s bound applies to any energy on a normalised state, not just the contractor's:

1. Extend the tier-7 variational bound to `compute_energy_ctm_tensor`, `compute_energy_split_ctm_tensor` and `compute_energy_split_ctm_tensor_2site`. A CTM energy below the ED ground state is unambiguously wrong, and this needs no new machinery.
2. Add a magnitude anchor on a *known* state rather than a random one, so "≈ 0" is caught rather than merely "below the bound". A product state with a hand-computable energy is enough to start.
3. Register whatever lands in the `core` bucket. #878 survived because the fermionic suite is `algorithm`/`slow` and does not run on a PR.

Worth stating the non-goal: tier 5's infinite-vs-torus comparison should stay unasserted. The point is not to force those two numbers to agree — it is that *some* assertion has to stand between a fermionic CTM energy and nonsense.

Refs #878, #667, #392.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

A data point from #996's witness calibration that belongs in this issue's record: on manifestly generic FermionParity states (random init + 2 SU steps, min λ > 4e-3, trace-normalized RDM with O(1) entries and 7.5e-3 off-diagonal structure at D=3), `compute_energy_split_ctm_tensor_2site` returns **E ≈ +6e-8 (D=2) / +4e-7 (D=3)** at χ=8 — six-plus orders below any generic expectation for `H = -t(c†c+h.c.) + V nn` on a random state, and the #854 PSD warning fires on the same readouts (smallest RDM eigenvalue −4.4e-6 of spectral radius). The two-site RDM machinery itself behaves sensibly on the same inputs (it resolved 10× gauge-displacement signals in #996's calibration), so the anomaly is specific to the energy contraction or its normalization, not the environment wholesale. Consistent with this issue's premise that the fermionic energy path is uncertified, and with the #995 audit's unadjudicated sign sites in the moves/projector stack. Fixture recipe to reproduce is in #996's calibration script description.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @yingjerkao.

## Post-#997 energy-trust verdict: the fermionic energy path is sound; validity is gated by environment quality, and the #854 warning is the gate

All runs on the #998 head (`3bbeaf7`, exact sweep), 2-site checkerboard, `mu=2V` PH-symmetric grafted onto `spinless_fermion_gate` (per-bond `-(mu/4)(n_i+n_j)`), 150 steps at dt=0.05 + 50 at dt=0.01, split-CTM chi=16. Every arm reports the manual `tr(rho H)` from both RDMs **and** `compute_energy_split_ctm_tensor_2site` on the same state and environments.

### 1. The earlier "E ≈ 1e-7 on generic states" anomaly is closed: it was input corruption, not the energy path

The identical probe (D=3, V=1, seed 1, 2 SU steps) post-fix reads `E_manual = +2.27e-02`, `E_fn = +2.68e-02` — five orders above the pre-fix reading. The pre-fix sweep had drained the state to near-vacuum (#997), where every t-V term vanishes identically; the energy path faithfully reported a corrupted state. Consistent with the sign census (0/4160 evaluations fire on the CTM+energy path).

### 2. On converged environments the two energy paths agree exactly, and hit the analytic target

D=2 CDW bench, both seeds:

| V | mu | ⟨n_A⟩/⟨n_B⟩ | E_manual/site | E_fn/site | analytic −V |
|---|---|---|---|---|---|
| 3 | 6 | 0.000 / 1.000 | −3.000000 | −3.000000 | −3 |
| 4 | 8 | 0.000 / 1.000 | −4.000000 | −4.000000 | −4 |

Six-decimal agreement with the pure-CDW analytic value, function ≡ manual, RDMs exactly pure `|01>`/`|10>` with the orientation flipping between seeds (spontaneous symmetry breaking, as it must). This is the May-2026 "strongest internal consistency check" reproduced on the **current** checkerboard sweep — the first time it holds post-#881, since the pre-#997 sweep destroyed the state before the energy ever saw it. The D=3 CDW-biased arm agrees between paths to 1e-3 at E=−2.72/site.

### 3. Where the paths disagree, the RDM is non-PSD and the #854 warning has already fired

Every arm with `E_fn` ≠ `E_manual` beyond ~1% is an arm whose RDM diagonal carries negative entries (up to −0.44) or entries >1 (up to +2.22): D=3 random seeds mid-evolution, D=2 V=0 free fermions at chi=16, the V=2 transition region. On those, manual and function disagree because **neither is an expectation value** — the #854 warning text says exactly this. The practical rule this suite establishes: **a fermionic energy is trustworthy iff the run is free of the #854 PSD warning**; energies from warned environments must be discarded, not averaged.

### 4. Physics notes from the scan (for #882 v2)

- V=1 (below the CDW transition) sits symmetric at E≈−1.500 — no charge order is correct there.
- V=2 (transition region) does not lock into the CDW within 200 steps from random D=2 seeds (E ≈ −1.83…−1.97 vs −2.0 pure-CDW); near-critical slowing, not a defect signal, but worth a longer-tau arm in any v2 acceptance sweep.
- D=3 attractor re-test: reversed — see the #882 comment; the "textbook SU pathology" verdict was an artifact of the #997 corruption.

### Disposition

With #998 merged, I consider the energy-trust question this issue asks **answered**: the contraction/normalization machinery is correct (analytic match + path agreement on healthy states), and the failure mode is environment quality, which is already self-flagging via #854. What remains open here is turning the rule in (3) into code — e.g. `fpeps()` refusing to report (or NaN-ing) an energy whose environment fired the PSD warning — plus the #995 projector-site adjudication.
